### 1. Dataset Definition
We define a simple linear dataset where the relationship is $y = 2x + 1$.

In [61]:
import numpy as np

X = np.array([1, 2, 3, 4, 5, 6, 7], dtype=float)

y = np.array([3, 5, 7, 9, 11, 13, 15], dtype=float)

### 2. Hyperparameters and Initialization
We initialize the weight $w$ and bias $b$ to $0.0$, and configure the learning rate and number of epochs.

In [62]:
w = 0.0
b = 0.0

learning_rate = 0.0001
epochs = 100000

### 3. Gradient Descent Training
We train the model using gradient descent to minimize Mean Squared Error (MSE).

In [63]:
for i in range(epochs):
    y_pred = w * X + b
    loss = np.mean((y_pred - y) ** 2)

    dw = (1 / len(X)) * np.sum((y_pred - y) * X)
    db = (1 / len(X)) * np.sum(y_pred - y)

    w = w- learning_rate * dw
    b = b- learning_rate * db


print(loss, w, b)

0.001404108002581948 2.0169168559755297 0.9162287270328173


### 4. Prediction
Using our trained parameters $w \approx 2.02$ and $b \approx 0.92$, we make a prediction for $x = 8$ (Expected: $17$).

In [64]:
new_x = 8

y_test = w * new_x + b

print(y_test)

17.051563574837054
